lib/phlex/elements.rb

```diff
@@ -20,11 +20,11 @@ def register_element(element, tag: element.name.tr("_", "-"))
 			def __phlex_#{element}__(**attributes, &block)
 				if attributes.length > 0
 					if block_given?
-						@_target << "<#{tag}" << (Phlex::ATTRIBUTE_CACHE[attributes.hash] || __attributes__(**attributes)) << ">"
+						@_target << "<#{tag}" << (Phlex::ATTRIBUTE_CACHE[respond_to?(:process_attributes) ? (attributes.hash + self.class.hash) : attributes.hash] || __attributes__(**attributes)) << ">"
```

Comment:
I wonder how brutal the performance hit would be to use `unbounded_method(:process_attributes).hash` instead of the class's hash.

I think that would make it so the attribute cache is only busted if you redefine `process_attributes`.

I'll try and benchmark it later today if I get time.

Reply:
It's 3 times slower 😢

```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
Warming up --------------------------------------
self.class.instance_method.hash
                       682.322k i/100ms
     self.class.hash     1.605M i/100ms
Calculating -------------------------------------
self.class.instance_method.hash
                          9.102M (± 2.2%) i/s -     45.716M in   5.025309s
     self.class.hash     28.106M (± 1.9%) i/s -    141.246M in   5.027368s

Comparison:
     self.class.hash: 28106052.4 i/s
self.class.instance_method.hash:  9101860.8 i/s - 3.09x  (± 0.00) slower
```

I think basing the cache key off the method implementation is the more correct caching behavior, but since the cache is just there to be fast, I'm not sure that having ideal caching behavior at the expense of speed is a good trade off.

I don't intend to add a `process_attributes` method on my `ApplicationView`, so I'd be fine to see it ship like this. It would be worthwhile to point out in the documentation.

Reply:
Hmm, it doesn't appear to affect performance much though in the bench.rb file. I modified the Layout fixture to define `process_attributes` like this:

```ruby
def process_attributes(a)
  a.transform_keys(&:upcase)
end
```

And it's running at basically the same speed:

## Process Attributes defined on Layout + instance_method.hash
```
env RUBY_YJIT_ENABLE=1 ruby bench.rb
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
Warming up --------------------------------------
                Page     9.990k i/100ms
Calculating -------------------------------------
                Page    105.026k (± 2.3%) i/s -    529.470k in   5.044352s
```

## Process Attributes defined on Layout + class.hash
```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
Warming up --------------------------------------
                Page    10.719k i/100ms
Calculating -------------------------------------
                Page    104.551k (± 3.5%) i/s -    525.231k in   5.030099s
```

## Process Attributes _not_ defined, with instance_method.hash
```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
Warming up --------------------------------------
                Page    10.265k i/100ms
Calculating -------------------------------------
                Page    104.474k (± 2.4%) i/s -    523.515k in   5.014144s
```

## Process Attributes _not_ defined, with class.hash
```
ruby 3.2.1 (2023-02-08 revision 31819e82c8) +YJIT [arm64-darwin22]
Warming up --------------------------------------
                Page    10.338k i/100ms
Calculating -------------------------------------
                Page    103.636k (± 2.3%) i/s -    527.238k in   5.090344s
```

I would have expected a 3x difference in the microbenchmark to be more measurable in the macrobenchmark. But maybe it's not a huge penalty 🤷? Or maybe I tested it wrong 😬

Reply:
_Narrator: He tested it wrong._

I didn't notice that the `bench.rb` is requiring the phlex gem, and not the local relative phlex project. So none of my changes were being benchmarked 🙃 

It's about half as fast to check the method's hash than to check the class's hash. So I think class check is the way to go.

Reply:
You can run `bundle exec ./bench.rb` to use the bundled gems.